cpp/src/arrow/ipc/reader.h

```diff
@@ -190,6 +193,85 @@ class ARROW_EXPORT RecordBatchFileReader {
   virtual Status ReadRecordBatch(int i, std::shared_ptr<RecordBatch>* batch) = 0;
 };
 
+/// \class RecordBatchReceiverCollect
+/// \brief Collect record batches read by RecordBatchStreamEmitter.
+class ARROW_EXPORT RecordBatchReceiverCollect : public Receiver {
+ public:
+  RecordBatchReceiverCollect() : record_batches_() {}
+  virtual ~RecordBatchReceiverCollect() = default;
+
+  Status RecordBatchReceived(std::shared_ptr<RecordBatch> record_batch) override {
+    record_batches_.push_back(std::move(record_batch));
+    return Status::OK();
+  }
+
+  /// \return the all read record batches
+  std::vector<std::shared_ptr<RecordBatch>> record_batches() const {
+    return record_batches_;
+  }
+
+ private:
+  std::vector<std::shared_ptr<RecordBatch>> record_batches_;
+};
+
+/// \class RecordBatchStreamEmitter
+/// \brief Push style record batch stream reader that receives data
+/// from user.
+///
+/// This class reads the schema (plus any dictionaries) as the first
+/// messages in the fed data, followed by record batches.
+class ARROW_EXPORT RecordBatchStreamEmitter {
+ public:
+  /// \brief Construct a record batch emitter.
+  ///
+  /// \param[in] receiver a Receiver that must implement
+  /// Receiver::MessageReceived to receive read record batches
+  /// \param[in] options any IPC reading options (optional)
+  RecordBatchStreamEmitter(Receiver* receiver,
+                           const IpcReadOptions& options = IpcReadOptions::Defaults());
+
+  virtual ~RecordBatchStreamEmitter();
+
+  /// \brief Feed data to the emitter as a raw data.
+  ///
+  /// If the emitter can read one or more record batches by the data,
+  /// the emitter emits read record batches by calling
+  /// receiver->Receive() multiple times.
+  ///
+  /// \param[in] data a raw data to be processed. This data isn't
+  /// copied. The passed memory must be kept alive through record
+  /// batch processing.
+  /// \param[in] size raw data size.
+  /// \return Status
+  Status Consume(const uint8_t* data, int64_t size);
+
+  /// \brief Feed data to the emitter as a Buffer.
+  ///
+  /// If the emitter can read one or more record batches by the
+  /// Buffer, the emitter emits read record batches by calling
+  /// receiver->Receive() multiple times.
+  ///
+  /// \param[in] buffer a Buffer to be processed.
+  /// \return Status
+  Status Consume(std::shared_ptr<Buffer> buffer);
+
+  /// \return the shared schema of the record batches in the stream
+  std::shared_ptr<Schema> schema() const;
+
+  /// \return the number of bytes needed to run the next action
```

Comment:
I don't understand what this means. What is the next action?

Reply:
I think it's "advancing the state of the emitter". So if you just read the metadata size prefix then this would return the size of the metadata, or the size of the body

Reply:
Does it mean you're expected to give exactly that number of bytes to `Consume`? Or does the emitter do its own buffering inside? The docs should probably make that clear.

Reply:
If you feed the emitter too much data, it will retain a slice of it internally, yes. This can be made more clear in the docs indeed

Reply:
I'll add more documentations. Could you confirm it?